In [1]:
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import warnings
warnings.filterwarnings('ignore')

In [3]:
train = pd.read_csv('../Data/PJT002_train.csv')
test = pd.read_csv('../Data/PJT002_test.csv')
valid = pd.read_csv('../Data/PJT002_validation.csv')

print('train', train.shape)
print('test', test.shape)
print('valid', valid.shape)

train.sample()

train (59199, 172)
test (2957, 171)
valid (6898, 172)


id             dt_of_fr bldng_us bldng_archtctr  bldng_cnt  \
30117  120650  2015-04-01 16:03:16      NaN            NaN          2   

       bldng_ar  ttl_ar  lnd_ar dt_of_athrztn  ttl_grnd_flr  ...  mlt_us_yn  \
30117       0.0     0.0     0.0           NaN           NaN  ...          N   

      cctv_dstnc  cctv_in_100m  fr_wthr_fclt_in_100m  tbc_rtl_str_dstnc  \
30117        602             0                     0               5962   

       sft_emrgnc_bll_dstnc  ahsm_dstnc  no_tbc_zn_dstnc  bldng_cnt_in_50m  \
30117                 31499       26147             3427                 1   

       fr_yn  
30117      N  

[1 rows x 172 columns]

In [4]:
# train-test-valid 데이터셋 모두 합쳐서 전처리
train['tag'] = 'train'
test['tag'] = 'test'
valid['tag'] = 'valid'

df = pd.concat([train, test, valid])
df.sample()

ahsm_dstnc  bldng_ar  bldng_ar_prc bldng_archtctr  bldng_cnt  \
46319        4735       0.0      385135.0            NaN          1   

       bldng_cnt_in_50m bldng_us bldng_us_clssfctn  cctv_dstnc  cctv_in_100m  \
46319                 3      NaN               NaN        8551             0   

       ... rgnl_ar_nm2 sft_emrgnc_bll_dstnc    tag  tbc_rtl_str_dstnc  tmprtr  \
46319  ...      지정되지않음                84183  train                366    29.7   

       ttl_ar  ttl_dwn_flr  ttl_grnd_flr  wnd_drctn  wnd_spd  
46319     0.0          NaN           NaN      230.0      1.4  

[1 rows x 173 columns]

In [5]:
# 컬럼명 확인
pd.DataFrame(df.columns)

0
0              ahsm_dstnc
1                bldng_ar
2            bldng_ar_prc
3          bldng_archtctr
4               bldng_cnt
5        bldng_cnt_in_50m
6                bldng_us
7       bldng_us_clssfctn
8              cctv_dstnc
9            cctv_in_100m
10          dt_of_athrztn
11               dt_of_fr
12    ele_engry_us_201401
13    ele_engry_us_201402
14    ele_engry_us_201403
15    ele_engry_us_201404
16    ele_engry_us_201405
17    ele_engry_us_201406
18    ele_engry_us_201407
19    ele_engry_us_201408
20    ele_engry_us_201409
21    ele_engry_us_201410
22    ele_engry_us_201411
23    ele_engry_us_201412
24    ele_engry_us_201501
25    ele_engry_us_201502
26    ele_engry_us_201503
27    ele_engry_us_201504
28    ele_engry_us_201505
29    ele_engry_us_201506
..                    ...
143        lnd_us_sttn_nm
144           lw_13101010
145           lw_13101110
146           lw_13101210
147           lw_13101211
148           lw_13101310
149           lw_13101410
150           lw_13111010
151           lw_13111110
152           lw_13121010
153           lw_13121011
154           lw_13131010
155           lw_13131110
156           lw_13141010
157           lw_13141011
158             mlt_us_yn
159       no_tbc_zn_dstnc
160               prcpttn
161              rd_sd_nm
162            rgnl_ar_nm
163           rgnl_ar_nm2
164  sft_emrgnc_bll_dstnc
165                   tag
166     tbc_rtl_str_dstnc
167                tmprtr
168                ttl_ar
169           ttl_dwn_flr
170          ttl_grnd_flr
171             wnd_drctn
172               wnd_spd

[173 rows x 1 columns]

## Null 값 확인

In [6]:
train.isnull().sum()

id                          0
dt_of_fr                    0
bldng_us                27677
bldng_archtctr          27665
bldng_cnt                   0
bldng_ar                    0
ttl_ar                      0
lnd_ar                      0
dt_of_athrztn           27581
ttl_grnd_flr            10210
ttl_dwn_flr             11005
bldng_us_clssfctn       29343
tmprtr                     13
prcpttn                 53629
wnd_spd                    33
wnd_drctn                 221
hmdt                       22
gas_engry_us_201401     53825
ele_engry_us_201401     43550
gas_engry_us_201402     53825
ele_engry_us_201402     43550
gas_engry_us_201403     53825
ele_engry_us_201403     43550
gas_engry_us_201404     53825
ele_engry_us_201404     43550
gas_engry_us_201405     53825
ele_engry_us_201405     43550
gas_engry_us_201406     53825
ele_engry_us_201406     43550
gas_engry_us_201407     53825
                        ...  
lw_13111010             58899
lw_13111110             58918
lw_1312101

- ['bldng_us_clssfctn', 29343]건물용도분류명(주거용)   
- ['bldng_us', 27677]건물용도(단독주택)  
- ['bldng_archtctr', 27665]건물구조(벽돌구조)  
- ['dt_of_athrztn', 27581]건물승인일자(19850417)  
- ['bldng_ar_prc', 21895]단위 면적당 건물 가격_2019년 기준(312411)  
- ['ttl_dwn_flr', 11005]건물들의 지하 층수의 합(0)  
- ['ttl_grnd_flr', 10210]건물들의 지상 층수의 합(4)  
- ['lnd_us_sttn_nm', 1776]토지 이용상황명(전)
- ['rd_sd_nm', 1491]도로측면명(맹지)  
- ['rgnl_ar_nm2', 1491]용도지역지구명2(지정되지 않음)  
- ['rgnl_ar_nm', 1491]용도지역지구명(개발제한구역)  
- ['hm_cnt', 701]행정구역 인구(27312)  
- ['wnd_drctn', 221]풍향(270)
- ['fr_mn_cnt', 41]관할 소방서 인원(224)  
- ['wnd_spd', 33]풍속(2.2)
- ['hmdt', 22]습도(77)
- ['tmprtr', 13]온도_c(4.4)  
- ['emd_nm', 4]행정구역명(경상남도 진주시 판문동)  
__________________________________________________
- ['bldng_cnt_in_50m', 0]반경 50M 이내의 건물 수(23)  
- ['no_tbc_zn_dstnc', 0]금연구역과의 최소 거리(235)  
- ['fr_wthr_fclt_in_100m', 0]반경 100M 이내 소방용수 시설 수(2)  
- ['ahsm_dstnc', 0]자동 심장 충격기와의 최소 거리(234)  
- ['sft_emrgnc_bll_dstnc', 0]안전 비상벨과의 최소 거리(233)  
- ['tbc_rtl_str_dstnc', 0]담배 소매점과의 최소 거리(232)  
- ['cctv_in_100m', 0]반경 100M 이내 공공 CCTV(1)  
- ['mlt_us_yn', 0]다중이용시설 포함여부(Y)  
- ['cctv_dstnc', 0]공공 CCTV와의 최소 거리(232)  
- ['fr_wthr_fclt_dstnc', 0]소방용수시설(소화전 등)과의 최소 거리(223)  
- ['fr_sttn_dstnc', 0]119안전센터와의 거리(3222)  
- ['jmk', 0]지적상 지목(답)  
- [dt_of_fr', 0]화재발생일시(2014-03-18 15:23:18)  
- ['lnd_ar', 0]토지면적(390)  
- ['ttl_ar', 0]건물연면적_건물층별합계전체면적(130.845)  
- ['bldng_ar', 0]건물건축면적(130.845)  
- ['bldng_cnt', 0]건물채수(4)  

## 전기/가스 사용량 컬럼 제외
Null 값도 많고, 컬럼 수가 너무 많아서 제외

In [7]:
columns = ['id', 'dt_of_fr', 'bldng_us', 'bldng_archtctr', 'bldng_cnt'
           , 'bldng_ar', 'ttl_ar', 'lnd_ar', 'dt_of_athrztn', 'ttl_grnd_flr'
           , 'ttl_dwn_flr', 'bldng_us_clssfctn', 'tmprtr', 'prcpttn'
           , 'wnd_spd', 'wnd_drctn', 'hmdt'
           , 'jmk' 
           , 'emd_nm', 'hm_cnt'
           , 'fr_sttn_dstnc', 'fr_wthr_fclt_dstnc'
           , 'mlt_us_yn', 'cctv_dstnc', 'cctv_in_100m', 'fr_mn_cnt'
           , 'fr_wthr_fclt_in_100m', 'tbc_rtl_str_dstnc', 'sft_emrgnc_bll_dstnc'
           , 'ahsm_dstnc', 'no_tbc_zn_dstnc', 'bldng_cnt_in_50m', 'fr_yn'
          ]

df = train[columns]
df.sample()

id             dt_of_fr bldng_us bldng_archtctr  bldng_cnt  \
27083  78352  2017-07-10 19:14:40      NaN            NaN          1   

       bldng_ar  ttl_ar  lnd_ar dt_of_athrztn  ttl_grnd_flr  ...  cctv_dstnc  \
27083    115.62  324.57   268.3           NaN           2.0  ...          67   

      cctv_in_100m  fr_mn_cnt  fr_wthr_fclt_in_100m  tbc_rtl_str_dstnc  \
27083            2      176.0                     0                174   

       sft_emrgnc_bll_dstnc  ahsm_dstnc no_tbc_zn_dstnc bldng_cnt_in_50m  \
27083                 10662         246             154               13   

       fr_yn  
27083      N  

[1 rows x 33 columns]

# PreProcessing

## One Hot Encoding) fr_yn 화재발생여부
- Y: 1
- N: 0

In [8]:
df['fr_yn'].unique()

array(['N', 'Y'], dtype=object)

In [9]:
df['fire_yes'] = (df['fr_yn'] == 'Y')

df[['fr_yn', 'fire_yes']].sample()

fr_yn  fire_yes
14712     N     False

## One Hot Encoding) mlt_us_yn 다중이용시설 포함여부

In [10]:
df['mlt_us_yn'].unique()

array(['N', 'Y'], dtype=object)

In [11]:
df['mlt_us_yn(encode)'] = df['mlt_us_yn'] == 'Y'

df[['mlt_us_yn', 'mlt_us_yn(encode)']].sample()

mlt_us_yn  mlt_us_yn(encode)
40670         N              False

In [12]:
data_fire = pd.DataFrame(df.groupby('mlt_us_yn(encode)')['fire_yes'].sum()).reset_index()
data_all = pd.DataFrame(df.groupby('mlt_us_yn(encode)')['id'].count()).reset_index()
data = data_fire.merge(data_all, on='mlt_us_yn(encode)')
data['ratio'] = data['fire_yes'] * 100 / data['id']
data

mlt_us_yn(encode)  fire_yes     id      ratio
0              False    7548.0  59010  12.791052
1               True     109.0    189  57.671958

## datetime Parsing) dt_of_fr 화재발생시점/관찰시점

In [13]:
df['year'] = pd.to_datetime(df['dt_of_fr']).dt.year
df['month'] = pd.to_datetime(df['dt_of_fr']).dt.month
df['day'] = pd.to_datetime(df['dt_of_fr']).dt.day
df['hour'] = pd.to_datetime(df['dt_of_fr']).dt.hour
df['minute'] = pd.to_datetime(df['dt_of_fr']).dt.minute
df['second'] = pd.to_datetime(df['dt_of_fr']).dt.second
df['dayofweek'] = pd.to_datetime(df['dt_of_fr']).dt.day_name()

df['year-month'] = df['dt_of_fr'].apply(lambda x: x[:7])
df['year-month-day'] = df['dt_of_fr'].apply(lambda x: x[:10])

df[['dt_of_fr', 'year-month', 'year-month-day', 'year', 'month', 'day', 'dayofweek', 'hour', 'minute', 'second']].sample()

dt_of_fr year-month year-month-day  year  month  day  \
42137  2016-02-17 13:33:32    2016-02     2016-02-17  2016      2   17   

       dayofweek  hour  minute  second  
42137  Wednesday    13      33      32

## Null 값 채우기) emd_nm 행정구역명
전체 데이터 중 행정구역명이 Null 값인 데이터는 4개 뿐이라 삭제

In [14]:
df = df.dropna(subset=['emd_nm'])

## 행정구역명 나눠서 컬럼'address1', 'address2'에 담기

In [15]:
df['address1'] = np.nan
df['address2'] = np.nan

df['address1'] = df['emd_nm'].apply(lambda x: str(x).split()[1][:3])
df['address2'] = df['emd_nm'].apply(lambda x: str(x).split()[2])
df['address'] = df['address1'] + ' ' + df['address2']
# 창원시는 따로
df.loc[df['address1'].apply(lambda x: len(x) > 3), 'address2'] = df['emd_nm'].apply(lambda x: str(x).split()[1][3:])

df[['emd_nm', 'address', 'address1', 'address2']].sample()

emd_nm  address address1 address2
44314  경상남도 거창군 거창읍  거창군 거창읍      거창군      거창읍

## Null 값 채우기) hm_cnt 행정구역 인구 

인구 Null 값인 행정구역을 찾아보니, 몇 개 행정구역에 몰려있음

In [16]:
df[df['hm_cnt'].isnull()]['address'].value_counts()

진주시 상대동      245
진주시 하대동      167
창원시 마산회원구     96
창원시 마산합포구     80
함안군 칠원읍       59
창원시 진해구        9
통영시 산양읍        6
통영시 욕지면        6
사천시 동금동        3
거제시 거제면        2
하동군 금남면        2
통영시 용남면        2
창녕군 대합면        2
하동군 진교면        2
고성군 거류면        2
진주시 상평동        1
거제시 아주동        1
창녕군 이방면        1
고성군 회화면        1
사천시 실안동        1
남해군 이동면        1
양산시 하북면        1
남해군 고현면        1
통영시 도산면        1
남해군 남해읍        1
거제시 장목면        1
통영시 미수동        1
남해군 창선면        1
거제시 둔덕면        1
Name: address, dtype: int64

### 행정구역별로 뜯어보기

#### 진주시 인구 Null 값
> 진주시청 제공, 12월말 기준

진주시 동 | 2014 | 2015 | 2016
-|-|-|-
진주시 상대동| 26_019 | 25_599 | 25_321
진주시 하대동| 27_457 | 26_981 | 27_067


In [17]:
df[df['address'] == '진주시 상대동'].groupby('year')['hm_cnt'].median()

year
2014        NaN
2015        NaN
2016        NaN
2017    22764.0
2018    22296.0
Name: hm_cnt, dtype: float64

In [18]:
df[df['address'] == '진주시 하대동'].groupby('year')['hm_cnt'].median()

year
2014        NaN
2015        NaN
2016        NaN
2017    28687.5
2018    28164.0
Name: hm_cnt, dtype: float64

#### 창원시는 null 값 뿐임. 인구 정보 따로 찾아보기
> 창원시청 제공, 12월말 기준

창원시 구 | 2014 | 2015 | 2016 | 2017 | 2018
-|-|-|-|-|-
창원시 마산합포구 | 184_179 | 183_829 | 183_989 | 181_917 | 182_772
창원시 마산회원구 | 216_749 | 214_264 | 213_163 | 203_926 | 201_565
창원시 진해구 | - | 189_463 | 190_780 | - |196_665

In [19]:
df[df['address'] == '창원시 마산회원구'].groupby('year')['hm_cnt'].median()

year
2014   NaN
2015   NaN
2016   NaN
Name: hm_cnt, dtype: float64

In [20]:
df[df['address'] == '창원시 마산합포구'].groupby('year')['hm_cnt'].median()

year
2014   NaN
2015   NaN
2016   NaN
2017   NaN
Name: hm_cnt, dtype: float64

In [21]:
df[df['address'] == '창원시 진해구'].groupby('year')['hm_cnt'].median()

year
2015   NaN
2016   NaN
2018   NaN
Name: hm_cnt, dtype: float64

#### 함안군 칠원읍

In [22]:
df[df['address'] == '함안군 칠원읍'].groupby('year')['hm_cnt'].median()

year
2014        NaN
2015    21131.0
2016    21022.0
2017    20889.5
2018    20672.0
Name: hm_cnt, dtype: float64

#### 사천시 동금동, 사천시 실안동 동단위 자료 없음
사천시 전체에서 적은 비중이라 삭제

In [23]:
df[df['address'] == '사천시 동금동'].groupby('year')['hm_cnt'].median()

year
2014   NaN
2017   NaN
Name: hm_cnt, dtype: float64

In [24]:
df[df['address'] == '사천시 실안동'].groupby('year')['hm_cnt'].median()

year
2017   NaN
Name: hm_cnt, dtype: float64

In [25]:
df[df['address1'] == '사천시']['address'].value_counts()

사천시 사천읍     380
사천시 사남면     278
사천시 용현면     270
사천시 동서동     266
사천시 서포면     262
사천시 곤양면     243
사천시 남양동     239
사천시 정동면     235
사천시 벌룡동     234
사천시 향촌동     209
사천시 곤명면     205
사천시 선구동     146
사천시 동서금동    138
사천시 축동면     127
사천시 동금동       3
사천시 실안동       1
Name: address, dtype: int64

In [26]:
# id 113578, 118226, 104635
# index 20446 20673 20680
df[df['address'] == '사천시 동금동'] 

# id 92201
# index 20676
# df[df['address'] == '사천시 실안동'] 

id             dt_of_fr   bldng_us bldng_archtctr  bldng_cnt  \
20446  113578  2017-06-26 20:12:00  제1종근린생활시설       철근콘크리트구조          1   
20673  118226  2014-08-08 14:26:00  제1종근린생활시설       철근콘크리트구조          1   
20680  104635  2014-05-22 21:38:00  제1종근린생활시설       철근콘크리트구조          1   

       bldng_ar  ttl_ar  lnd_ar dt_of_athrztn  ttl_grnd_flr  ...  day hour  \
20446    225.21  203.73   675.0   2.00502e+07           1.0  ...   26   20   
20673    225.21  203.73   675.0   2.00502e+07           1.0  ...    8   14   
20680    225.21  203.73   675.0   2.00502e+07           1.0  ...   22   21   

       minute  second  dayofweek  year-month  year-month-day address1  \
20446      12       0     Monday     2017-06      2017-06-26      사천시   
20673      26       0     Friday     2014-08      2014-08-08      사천시   
20680      38       0   Thursday     2014-05      2014-05-22      사천시   

      address2  address  
20446      동금동  사천시 동금동  
20673      동금동  사천시 동금동  
20680      동금동  사천시 동금동  

[3 rows x 47 columns]

In [27]:
# 사천시 실안동, 동금동 row 삭제

df = df.drop([20446, 20673, 20680, 20676])

### 연도별 인구 중간값으로 채우기

In [64]:
df['hm_cnt'] = df.groupby(['year', 'address'])['hm_cnt'].transform(lambda x: x.fillna(x.median()))

In [65]:
# 남은 null 값 확인

df['hm_cnt'].isnull().sum()

185

In [79]:
df[df['hm_cnt'].isnull()].groupby(['address', 'year'])['hm_cnt'].size()

address    year
창원시 마산합포구  2014    30
           2015    24
           2016    25
           2017     1
창원시 마산회원구  2014    34
           2015    23
           2016    39
창원시 진해구    2015     2
           2016     3
           2018     4
Name: hm_cnt, dtype: int64

In [105]:
마산회원구 = {2014: 216_749, 2015: 214_264, 2016: 213_163, 2017: 203_926, 2018: 201_565}
마산합포구 = {2014: 184_179, 2015: 183_829, 2016: 183_989, 2017: 181_917, 2018: 182_772}
진해구 = {2014: np.nan, 2015: 189_463, 2016: 190_780, 2017: np.nan, 2018: 196_665}

hm_cnt = pd.DataFrame(data=마산회원구.values(), index=마산회원구.keys(), columns=['마산회원구'])
hm_cnt['마산합포구'] = 마산합포구.values()
hm_cnt['진해구'] = 진해구.values()

hm_cnt

마산회원구   마산합포구     진해구
2014  216749  184179       0
2015  214264  183829  189463
2016  213163  183989  190780
2017  203926  181917       0
2018  201565  182772  196665

In [107]:
# 창원시 인구 Null 값 채우기
for place in hm_cnt.columns:
    for year in hm_cnt.index:
        df.loc[df['address2'] == place, 'hm_cnt'] = hm_cnt[place][year]

In [108]:
# Null 값 잘 채워졌는지 확인

df['hm_cnt'].isnull().sum()

0

## Null 값 채우기) fr_mn_cnt 관할소방서 인원 

In [109]:
df['fr_mn_cnt'] = df.groupby(['year', 'address'])['fr_mn_cnt'].transform(lambda x: x.fillna(x.median()))

In [112]:
# 소방서 인원 Null 값 채워졌는지 확인

df['fr_mn_cnt'].isnull().sum()

0

## Null 값 채우기) wnd_drctn 풍향
1. 해당 지역, 해당일의 풍향 중간값으로 채우기
2. 1에서 안 채워진 경우, 해당 지역, 해당월의 풍향 중간값으로 채우기
3. 2에서 안 채워진 경우, 해당일의 풍향 중간값으로 채우기

In [44]:
df['wnd_drctn'] = df.groupby(['address', 'year-month-day'])['wnd_drctn'].transform(lambda x: x.fillna(x.median()))

In [45]:
df['wnd_drctn'] = df.groupby(['address', 'year-month'])['wnd_drctn'].transform(lambda x: x.fillna(x.median()))

In [46]:
df['wnd_drctn'] = df.groupby(['year-month-day'])['wnd_drctn'].transform(lambda x: x.fillna(x.median()))

In [47]:
# 다 채워졌는지 확인
df['wnd_drctn'].isnull().sum()

0

## Null 값 채우기) hmdt 습도
corr 으로 습도와 상관계수 높은 컬럼 찾기

In [48]:
corr = train.corr()

In [49]:
corr.loc[corr['hmdt']<-0.2, 'hmdt']

wnd_spd     -0.402712
wnd_drctn   -0.253173
Name: hmdt, dtype: float64

1. 같은 풍속, 풍향인 날의 습도 중간값으로 채우기
2. 1에서 안 채워진 경우, 풍향이 같은 날의 습도 중간값으로 채우기 
3. 2에서 안 채워진 경우, 해당일의 습도 중간값으로 채우기

In [114]:
df['hmdt'].shape

(59191,)

In [115]:
df.shape

(59191, 49)

In [117]:
# 에러 Length Mismatch
df['hmdt'] = df.groupby(['wnd_spd', 'wnd_drctn'])['hmdt'].transform(lambda x: x.fillna(x.median()))

ValueError: Length mismatch: Expected axis has 59158 elements, new values have 59191 elements

In [51]:
df['hmdt'] = df.groupby(['year-month-day'])['hmdt'].transform(lambda x: x.fillna(x.median()))

In [52]:
df['hmdt'].isnull().sum()

0

## Null 값 채우기) tmprtr 온도
corr 으로 온도와 상관계수가 높은 컬럼 찾기

In [53]:
corr.loc[corr['tmprtr'] > 0.1, 'tmprtr']

tmprtr         1.000000
wnd_spd        0.106366
hmdt           0.125568
lw_13101110    0.145238
lw_13101210    0.118775
lw_13101211    0.136644
lw_13111010    0.116014
Name: tmprtr, dtype: float64

1. 풍속과 습도가 같은 날의 온도 중간값으로 채우기
2. 습도가 같은 해당일의 온도 중간값으로 채우기
3. 풍속이 같은 해당일의 온도 중간값으로 채우기
4. 습도가 같은 해당월의 온도 중간값으로 채우기

In [54]:
# 에러 Length Mismatch
df['tmprtr'] = df.groupby(['wnd_spd', 'hmdt'])['tmprtr'].transform(lambda x: x.fillna(x.median()))

ValueError: Length mismatch: Expected axis has 59158 elements, new values have 59191 elements

In [55]:
df['tmprtr'] = df.groupby(['hmdt', 'year-month-day'])['tmprtr'].transform(lambda x: x.fillna(x.median()))

In [56]:
# 에러 Length Mismatch
df['tmprtr'] = df.groupby(['wnd_spd', 'year-month-day'])['tmprtr'].transform(lambda x: x.fillna(x.median()))

ValueError: Length mismatch: Expected axis has 59158 elements, new values have 59191 elements

In [57]:
df['tmprtr'] = df.groupby(['hmdt', 'year-month'])['tmprtr'].transform(lambda x: x.fillna(x.median()))

In [58]:
# Null 값 잘 채워졌는지 확인
train['tmprtr'].isnull().sum()

13